### PR TITLE
ramips: Add support to TP-Link Archer MR200

### PR DESCRIPTION
**Pull request by heinzek**
- CPU: MT7620N 580MHz
- Flash: 8MB
- RAM: 64MB
- build-in modem 3G/4G
- four ethernet port 10/100Mbps
  Signed-off-by: Henryk Heisig hyniu@o2.pl

### CODE REVIEW

File: target/linux/ramips/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -201,6 +201,9 @@ ramips_setup_interfaces()
 	wl-330n)
 		ucidef_set_interface_lan "eth0.1"
 		;;
+	mr200)
+		ucidef_set_interfaces_lan_wan "eth0" "usb0"
+		;;
```
Comment (blogic) [inline target/linux/ramips/base-files/etc/board.d/02_network]: is there really a usb device on wan per default ?

File: target/linux/ramips/base-files/etc/init.d/enablemodem [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+. /lib/ramips.sh
+
+START=99
+
+start() {
+	local board=$(ramips_board_name)
+	if [ $board = "mr200" ]; then
+		adb shell chmod +x /WEBSERVER/www/cgi-bin/*
+		adb shell httpd -h /WEBSERVER/www/ &
+		echo "2357 000d" > /sys/bus/usb-serial/drivers/option1/new_id
```
Comment (blogic) [inline target/linux/ramips/base-files/etc/init.d/enablemodem]: adb is not part of the base packages

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 1 above)
Comment (heinzek) [inline target/linux/ramips/base-files/etc/board.d/02_network]: Yes. Modem use kmod-usb-net-rndis.

**Comment by blogic** [comment]
the modem runs on android ?!

**Comment by ignisf** [comment]
flash it with LEDE, we need to go deeper!

**Comment by heinzek** [comment]
Yes. Lede see: cat /sys/kernel/debug/usb/devices

T:  Bus=02 Lev=00 Prnt=00 Port=00 Cnt=00 Dev#=  1 Spd=12   MxCh= 1
B:  Alloc=  0/900 us ( 0%), #Int=  0, #Iso=  0
D:  Ver= 1.10 Cls=09(hub  ) Sub=00 Prot=00 MxPS=64 #Cfgs=  1
P:  Vendor=1d6b ProdID=0001 Rev= 3.18
S:  Manufacturer=Linux 3.18.23 ohci_hcd
S:  Product=Generic Platform OHCI controller
S:  SerialNumber=101c1000.ohci
C:* #Ifs= 1 Cfg#= 1 Atr=e0 MxPwr=  0mA
I:* If#= 0 Alt= 0 #EPs= 1 Cls=09(hub  ) Sub=00 Prot=00 Driver=hub
E:  Ad=81(I) Atr=03(Int.) MxPS=   2 Ivl=255ms

T:  Bus=01 Lev=00 Prnt=00 Port=00 Cnt=00 Dev#=  1 Spd=480  MxCh= 1
B:  Alloc=  0/800 us ( 0%), #Int=  1, #Iso=  0
D:  Ver= 2.00 Cls=09(hub  ) Sub=00 Prot=00 MxPS=64 #Cfgs=  1
P:  Vendor=1d6b ProdID=0002 Rev= 3.18
S:  Manufacturer=Linux 3.18.23 ehci_hcd
S:  Product=EHCI Host Controller
S:  SerialNumber=101c0000.ehci
C:* #Ifs= 1 Cfg#= 1 Atr=e0 MxPwr=  0mA
I:* If#= 0 Alt= 0 #EPs= 1 Cls=09(hub  ) Sub=00 Prot=00 Driver=hub
E:  Ad=81(I) Atr=03(Int.) MxPS=   4 Ivl=256ms

T:  Bus=01 Lev=01 Prnt=01 Port=00 Cnt=01 Dev#=  2 Spd=480  MxCh= 0
D:  Ver= 2.01 Cls=00(>ifc ) Sub=00 Prot=00 MxPS=64 #Cfgs=  1
P:  Vendor=2357 ProdID=000d Rev= 2.28
S:  Manufacturer=TP-LINK Technologies Co., Ltd.
S:  Product=MR200
S:  SerialNumber=b8135db25416
C:* #Ifs= 7 Cfg#= 1 Atr=a0 MxPwr=500mA
A:  FirstIf#= 0 IfCount= 2 Cls=e0(wlcon) Sub=01 Prot=03
I:* If#= 0 Alt= 0 #EPs= 1 Cls=e0(wlcon) Sub=01 Prot=03 Driver=rndis_host
E:  Ad=82(I) Atr=03(Int.) MxPS=   8 Ivl=32ms
I:* If#= 1 Alt= 0 #EPs= 2 Cls=0a(data ) Sub=00 Prot=00 Driver=rndis_host
E:  Ad=81(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=01(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 2 Alt= 0 #EPs= 2 Cls=ff(vend.) Sub=ff Prot=ff Driver=option
E:  Ad=83(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=02(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 3 Alt= 0 #EPs= 2 Cls=ff(vend.) Sub=42 Prot=01 Driver=option
E:  Ad=84(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=03(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 4 Alt= 0 #EPs= 3 Cls=ff(vend.) Sub=00 Prot=00 Driver=option
E:  Ad=86(I) Atr=03(Int.) MxPS=  10 Ivl=32ms
E:  Ad=85(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=04(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 5 Alt= 0 #EPs= 3 Cls=ff(vend.) Sub=00 Prot=00 Driver=option
E:  Ad=88(I) Atr=03(Int.) MxPS=  10 Ivl=32ms
E:  Ad=87(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=05(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
I:* If#= 6 Alt= 0 #EPs= 2 Cls=08(stor.) Sub=06 Prot=50 Driver=usb-storage
E:  Ad=89(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=06(O) Atr=02(Bulk) MxPS= 512 Ivl=125us


Modem have own webinterface, but it is not running. Packages adb is needed to enable it.

**Comment by bobafetthotmail** [comment]
Woho, that's great. That's great hardware.

>the modem runs on android ?!

many huawei usb dongles run on some kind of crappy embedded outdated android actually.
http://www.gnuton.org/blog/2015/07/huawei-e3372/
the e3372 for example 
_The firmware consists of two parts: WebUI and the main firmware. The **main firmware is based on Android 2.3** and runs a 3.4.5 linux kernel._

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 1 above)
Comment (mkresin) [inline target/linux/ramips/base-files/etc/board.d/02_network]: Please add a switch config here for the board. You are relying on settings done in [target/linux/ramips/base-files/lib/preinit/07_set_preinit_iface_ramips](https://github.com/lede-project/source/blob/40bfe9f0bdb4750a77140d1e402e9cf37326c38f/target/linux/ramips/base-files/lib/preinit/07_set_preinit_iface_ramips).

As soon as these settings are changed for some reason, networking for your board might not work anymore. You can set your wan interface using ```ucidef_set_interfaces_wan``` when having the switch config.

**Comment by bobafetthotmail** [comment]
>Modem have own webinterface, but it is not running. Packages adb is needed to enable it.

can the modem be used without its webinterface? did you try to install  drivers for 3g/4g modems?

these entries should show the driver name instead of "option" once you installed the drivers.
_I:* If#= 2 Alt= 0 #EPs= 2 Cls=ff(vend.) Sub=ff Prot=ff Driver=option

I:* If#= 3 Alt= 0 #EPs= 2 Cls=ff(vend.) Sub=42 Prot=01 Driver=option

I:* If#= 4 Alt= 0 #EPs= 3 Cls=ff(vend.) Sub=00 Prot=00 Driver=option

I:* If#= 5 Alt= 0 #EPs= 3 Cls=ff(vend.) Sub=00 Prot=00 Driver=option_

**Comment by heinzek** [comment]
Switch support is added.
What can I do with a package adb? 
This modem support AT command on /dev/ttyUSB1 but cant set property connection to internet.
If use modem gui to setup connection to internet, modem will automatically connect to the Internet. No adb and AT command needed.

**Comment by bobafetthotmail** [comment]
>If use modem gui to setup connection to internet, modem will automatically connect to the Internet. No adb and AT command needed.

Yes I know, these things are %&£%"£ to set up through serial.
But most of these modems in web-gui mode add a NAT and sometimes a firewall.
So features like dynamic DNS, opening ports towards Internet, VPN don't work.

**Comment by heinzek** [comment]
@bobafetthotmail This modem on default have configured DMZ to IP 192.168.225.100 When router have this IP on WAN all features working corectly.

**Comment by bobafetthotmail** [comment]
>@bobafetthotmail This modem on default have configured DMZ to IP 192.168.225.100 When router have this IP on WAN all features working corectly.

Good, thanks for checking. :+1:

**Comment by blogic** [comment]
for my undersdtanding there are 3 waays to do 3G uplink
1) via adb
2) via the webui
3) via AT calls (but this does not work properly)
is that correct ?

**Comment by heinzek** [comment]
No. Adb is only needed to enable and start web gui. Connection can be set property only in modem web gui.
On AT command can view only some Info about current connection like signall strenght or modem sn.

**Comment by HRomie** [comment]
FYI, i was able to successfully enable the LTE web gui:
https://forum.openwrt.org/viewtopic.php?pid=339205#p339205

File: target/linux/ramips/base-files/etc/init.d/enablemodem (hunk 2 above)
Comment (HRomie) [inline target/linux/ramips/base-files/etc/init.d/enablemodem]: the last line (the echo line), will make further connection the the LTE device through adb impossible. adb exits with this error: "error: device not found"

File: tools/firmware-utils/src/mktplinkfw2.c [hunk 3]
```diff
@@ -156,6 +156,12 @@ static struct flash_layout layouts[] = {
 		.kernel_ep	= 0x80000000,
 		.rootfs_ofs	= 0x140000,
 	}, {
+		.id		= "8MLmtk",
+		.fw_max_len	= 0x7b0000,
```
Comment (HRomie) [inline tools/firmware-utils/src/mktplinkfw2.c]: Should .fw_max_len be 0x7A0000 ?
Detail: https://forum.openwrt.org/viewtopic.php?pid=339116#p339116

tl;dr: if the final sysupgrade size goes beyond 0x7A0000 bytes, it's likely can not be used to initially switch to lede/openwrt via bootloaders tftp recovery (which is the only known method). However, it still could be used to upgrade the existing openwrt/lede installation.

File: target/linux/ramips/base-files/etc/init.d/enablemodem (hunk 2 above)
Comment (heinzek) [inline target/linux/ramips/base-files/etc/init.d/enablemodem]: @HRomie if You first run adb server then add new id to option driver You can connect to adb shell and use port tty to comunicate with modem.
Modem have 7 interfaces:
0 - rndis
1 - rndis
2 - option
3 - usbfs (adb interface)
4 - option
5 - option
6 - usbstorage

When You first use echo command then interface 3 use option driver.

File: tools/firmware-utils/src/mktplinkfw2.c (hunk 3 above)
Comment (HRomie) [inline tools/firmware-utils/src/mktplinkfw2.c]: At it's current form, sysupgrade does not appear to be able to keep the settings between each flash. the overlay partition keeps getting re-built after sysupgrade restart.

File: target/linux/ramips/base-files/etc/board.d/02_network [hunk 4]
```diff
@@ -201,6 +201,11 @@ ramips_setup_interfaces()
 	wl-330n)
 		ucidef_set_interface_lan "eth0.1"
 		;;
+	mr200)
+		ucidef_add_switch "switch0" \
+			"0:lan" "1:lan" "2:lan" "3:lan" "6t@eth0"
+		ucidef_set_interface_wan "usb0"
+		;;
```
Comment (HRomie) [inline target/linux/ramips/base-files/etc/board.d/02_network]: This is not working. the generated network file for wan, ends up being like this:

```
config interface 'wan'                    
        option ifname 'eth0.2'                 
        option proto 'dhcp'               
                                      
config device 'wan_dev'       
        option name 'eth0.2'              
        option macaddr '3c:46:d8:xx:xx:xx'
                                          
config interface 'wan6'                   
        option ifname 'eth0.2'            
        option proto 'dhcpv6'          
```

File: target/linux/ramips/base-files/etc/init.d/enablemodem (hunk 2 above)
Comment (HRomie) [inline target/linux/ramips/base-files/etc/init.d/enablemodem]: @heinzek
Confirmed. running the echo command after running the adb daemon, does not block the adb interface.
On a side note, this new file(enablemodem), must have +x permission set. i don't know how/where you define that in github but it ended up without +x in my build which prevented the router from running it.

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 4 above)
Comment (HRomie) [inline target/linux/ramips/base-files/etc/board.d/02_network]: We also likely need to manually specify usb0 ip address to 192.168.225.100/24 instead of relying on dhcp. DMZ is set on that address and the modem's dhcp server does not seem to be giving that address to the router as it's first choice. And of course if we do so, we also have to manually specify the gateway and dns.
what are your thoughts on this?

**Comment by heinzek** [comment]
@HRomie i update my branch now You can try build firmware.
Use: git clone https://github.com/heinzek/source.git
git checkout mr200
make menuconfig  (select target)
make

now im working to set static IP on wan interface

File: package/utils/adb/Makefile [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_SOURCE_VERSION)
+PKG_SOURCE:=$(PKG_SOURCE_SUBDIR).tar.gz
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_SOURCE_SUBDIR)
+PKG_MAINTAINER:=Swift Geek <swiftgeek at gmail>
+
+include $(INCLUDE_DIR)/package.mk
+
+ifeq ($(CONFIG_BIG_ENDIAN),y)
+TARGET_CFLAGS+= -DHAVE_BIG_ENDIAN=1
+beinfo: $(info *** Big endian detected, look for -DHAVE_BIG_ENDIAN=1 ***)
```
Comment (nbd168) [inline package/utils/adb/Makefile]: Unnecessary debug code, should be removed

File: package/utils/adb/Makefile [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+endif
+TARGET_CFLAGS+= -D_GNU_SOURCE
+
+define Package/adb
+	SECTION:=utils
+	CATEGORY:=Utilities
+	TITLE:=Android Debug Bridge CLI tool
+	URL:=http://tools.android.com/
+	DEPENDS:=+zlib +libopenssl +libpthread
```
Comment (nbd168) [inline package/utils/adb/Makefile]: Please use whitespace indentation instead of tabs here.

File: package/utils/adb/Makefile [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+define Package/bridge/description
+	Android Debug Bridge (adb) is a versatile command line tool that lets you communicate with an emulator instance or connected Android-powered device.
+endef
+
+# Add Makefile from AUR
+define Build/Prepare
+	$(call Build/Prepare/Default)
+	$(CP) ./adbMakefile $(PKG_BUILD_DIR)/adb/Makefile
+endef
```
Comment (nbd168) [inline package/utils/adb/Makefile]: You can remove the Build/Prepare override if you rename ./adbMakefile to ./files/adb/Makefile

File: package/utils/adb/Makefile [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+define Build/Configure
+endef
+
+define Build/Compile
+	$(MAKE) -C $(PKG_BUILD_DIR)/adb/ \
+		$(TARGET_CONFIGURE_OPTS) \
+		TARGET=Linux \
+		CFLAGS="$(TARGET_CFLAGS)" \
+		LDFLAGS="$(TARGET_LDFLAGS)"
+	$(STRIP) $(PKG_BUILD_DIR)/adb/adb
```
Comment (nbd168) [inline package/utils/adb/Makefile]: Please remove the $(STRIP) call. The build system takes care of that.

File: package/utils/adb/Makefile [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://android.googlesource.com/platform/system/core
+PKG_SOURCE_VERSION:=6fe92d1a3fb17545d82d020a3c995f32e6b71f9d
+# PKG_SOURCE_SUBDIR specifies where repo is cloned
+PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_SOURCE_VERSION)
+PKG_SOURCE:=$(PKG_SOURCE_SUBDIR).tar.gz
+PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_SOURCE_SUBDIR)
+PKG_MAINTAINER:=Swift Geek <swiftgeek at gmail>
```
Comment (nbd168) [inline package/utils/adb/Makefile]: Different maintainer than Signed-off-by?

File: package/utils/adb/Makefile [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+PKG_NAME:=adb
+PKG_VERSION:=android.5.0.2_r1
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://android.googlesource.com/platform/system/core
+PKG_SOURCE_VERSION:=6fe92d1a3fb17545d82d020a3c995f32e6b71f9d
+# PKG_SOURCE_SUBDIR specifies where repo is cloned
+PKG_SOURCE_SUBDIR:=$(PKG_NAME)-$(PKG_SOURCE_VERSION)
+PKG_SOURCE:=$(PKG_SOURCE_SUBDIR).tar.gz
```
Comment (nbd168) [inline package/utils/adb/Makefile]: Please use .xz

File: target/linux/ramips/dts/ArcherMR200.dts [hunk 11]
```diff
@@ -0,0 +1,193 @@
+/dts-v1/;
+
+#include "mt7620a.dtsi"
```
Comment (mkresin) [inline target/linux/ramips/dts/ArcherMR200.dts]: Please include <[dt-bindings/input/input.h](http://lxr.free-electrons.com/source/include/dt-bindings/input/linux-event-codes.h?v=4.4)> and  <[dt-bindings/gpio/gpio.h](http://lxr.free-electrons.com/source/include/dt-bindings/gpio/gpio.h?v=4.4)> here as well.

Use the constants GPIO_ACTIVE_LOW and GPIO_ACTIVE_HIGH afterwards instead of 1 and 0 in the gpio parameters.

Use the Key constants instead of the hex values in linux,code.

File: package/utils/adb/Makefile (hunk 8 above)
Comment (heinzek) [inline package/utils/adb/Makefile]: When I remove $strip, have error while compilation:
```
make[3]: Leaving directory '/home/hyniu/LEDE/build_dir/target-mipsel_24kc_musl-1.1.15/adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d/adb'
/home/hyniu/LEDE/build_dir/target-mipsel_24kc_musl-1.1.15/adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d/adb/adb
bash: /home/hyniu/LEDE/build_dir/target-mipsel_24kc_musl-1.1.15/adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d/adb/adb: cannot execute binary file: Exec format error
Makefile:53: recipe for target '/home/hyniu/LEDE/build_dir/target-mipsel_24kc_musl-1.1.15/adb-6fe92d1a3fb17545d82d020a3c995f32e6b71f9d/.built' failed

```
with this option compilation is without errors

File: package/utils/adb/Makefile (hunk 7 above)
Comment (heinzek) [inline package/utils/adb/Makefile]: Copy this adbMakefile to files directory not work, because files is copied to package after compile. I make patch to create this file.

File: target/linux/ramips/dts/ArcherMR200.dts (hunk 11 above)
Comment (heinzek) [inline target/linux/ramips/dts/ArcherMR200.dts]: fixed

File: package/utils/adb/Makefile (hunk 8 above)
Comment (heinzek) [inline package/utils/adb/Makefile]: fixed

File: target/linux/ramips/base-files/etc/board.d/02_network (hunk 4 above)
Comment (heinzek) [inline target/linux/ramips/base-files/etc/board.d/02_network]: On my mr200 this config always  work good: 
```
config interface 'loopback'
        option ifname 'lo'
        option proto 'static'
        option ipaddr '127.0.0.1'
        option netmask '255.0.0.0'

config globals 'globals'
        option ula_prefix 'fd1f:2c01:2cde::/48'

config interface 'lan'
        option type 'bridge'
        option ifname 'eth0.1'
        option proto 'static'
        option ipaddr '192.168.1.1'
        option netmask '255.255.255.0'
        option ip6assign '60'

config device 'lan_dev'
        option name 'eth0.1'
        option macaddr '3c:46:d8:aa:89:4f'

config interface 'wan'
        option ifname 'usb0'
        option proto 'dhcp'

config device 'wan_dev'
        option name 'usb0'
        option macaddr '3c:46:d8:aa:89:50'

config interface 'wan6'
        option ifname 'usb0'
        option proto 'dhcpv6'

config switch
        option name 'switch0'
        option reset '1'
        option enable_vlan '1'

config switch_vlan
        option device 'switch0'
        option vlan '1'
        option ports '0 1 2 3 6t'

```

But i dont know how to set static IP on usb0 interface with ip 192.168.225.100, mask 255.255.255.0 gateway 192.168.225.1 and dns 192.168.225.1

Comment (HRomie) [inline target/linux/ramips/base-files/etc/board.d/02_network]: Unfortunately I am using this modem in production now and so can not easily use it to re-test the settings. it is however possible that i've made a mistake the first time and the default configuration works correctly. If it works in your case, i'm sure it should be fine.

**Comment by heinzek** [comment]
@nbd168 everything ok on this PR?

**Comment by blogic** [comment]
Hi

patch 1/3 has pending review, please address this
patch 2/3 looks good
patch 3/3 there is not need for a dedicated package. put the script into target/linux/ramips/base-files/....

File: package/utils/adb/Makefile (hunk 6 above)
Comment (heinzek) [inline package/utils/adb/Makefile]: fixed

File: package/utils/adb/Makefile (hunk 9 above)
Comment (heinzek) [inline package/utils/adb/Makefile]: fixed

File: package/utils/adb/Makefile (hunk 10 above)
Comment (heinzek) [inline package/utils/adb/Makefile]: fixed

**Comment by heinzek** [comment]
@blogic patch 3 is merged to 1
Thanks for Your review

**Comment by HRomie** [comment]
@maxcherry89 
first read the openwrt wiki: https://wiki.openwrt.org/toh/tp-link/archer-mr200
then ask any question regarding it in the openwrt forum: http://forum.openwrt.org/viewtopic.php?id=64293

Basically yes, it is possible now to run LEDE on this device. and if we get this pull approved, it gets way easier as you wouldn't need to compile the firmware yourself.

What's holding this pull request by the way? i'm not quite familiar with github process for accepting a pull request but it does seem heinzek address most (if not all) of the concerns.

**Comment by blogic** [comment]
@HRomie its on my todo list for today ;)

**Comment by blogic** [comment]
merged, Thanks !
